### PR TITLE
feat: implement optimized vector of CIDs

### PR DESCRIPTION
**Pull request by jdjaustin**
## Summary of changes

Changes introduced in this pull request:

Implements a `CidVec` type similar to the optimized versions of `HashSet<Cid>` and `HashMap<Cid, _>`. Subtasks:
- [x] Implement `CidVec`.
- [x] Use `CidVec` in `TipsetKeys`.
- [x] Write quickcheck properties to make sure `CidVec` is identical to `Vec<Cid>`.

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/blocks/tipset.rs [hunk 1]
```diff
@@ -712,10 +714,10 @@ mod test {
             .unwrap();
         let h1 = BlockHeader::builder()
             .miner_address(Address::new_id(1))
-            .parents(TipsetKeys::new(vec![Cid::new_v1(
+            .parents(TipsetKeys::new(CidVec::new_from_cid(Cid::new_v1(
```
Comment (lemmih) [inline src/blocks/tipset.rs:717]: Let's add a `From<Cid>` implementation.

File: src/chain_sync/chain_muxer.rs [hunk 2]
```diff
@@ -375,7 +375,7 @@ where
                 metrics::LIBP2P_MESSAGE_TOTAL
                     .with_label_values(&[metrics::values::HELLO_RESPONSE_OUTBOUND])
                     .inc();
-                let tipset_keys = TipsetKeys::new(request.heaviest_tip_set);
+                let tipset_keys = TipsetKeys::new(request.heaviest_tip_set.into());
```
Comment (lemmih) [inline src/chain_sync/chain_muxer.rs:378]: Let's add a `From<Vec<Cid>>` implementation.

File: src/chain_sync/tipset_syncer.rs [hunk 3]
```diff
@@ -882,7 +882,7 @@ async fn sync_headers_in_reverse<DB: Blockstore + Clone + Sync + Send + 'static,
         }
         // Attempt to load the parent tipset from local store
         if let Ok(tipset) = chain_store.tipset_from_keys(oldest_parent.parents()) {
-            parent_blocks.extend_from_slice(tipset.cids());
+            parent_blocks.extend(tipset.cids());
```
Comment (lemmih) [inline src/chain_sync/tipset_syncer.rs:885]: Let's write an `IntoIterator<Item=Cid>` implementation:

```suggestion
            parent_blocks.extend(tipset);
```

File: src/cli/subcommands/chain_cmd.rs [hunk 4]
```diff
@@ -92,7 +92,9 @@ impl ChainCommands {
             } => {
                 maybe_confirm(*no_confirm, SET_HEAD_CONFIRMATION_MESSAGE)?;
                 chain_set_head(
-                    (TipsetKeys { cids: cids.clone() },),
+                    (TipsetKeys {
+                        cids: cids.clone().into(),
+                    },),
```
Comment (lemmih) [inline src/cli/subcommands/chain_cmd.rs:97]:
```suggestion
                    TipsetKeys::from(cids.clone()),
```

File: src/db/car/forest.rs [hunk 5]
```diff
@@ -136,7 +136,7 @@ impl<ReaderT: super::CarReader> ForestCar<ReaderT> {
     }
 
     pub fn heaviest_tipset(&self) -> anyhow::Result<Tipset> {
-        Tipset::load_required(self, &TipsetKeys::new(self.roots()))
+        Tipset::load_required(self, &TipsetKeys::new(self.roots().into()))
```
Comment (lemmih) [inline src/db/car/forest.rs:139]:
```suggestion
        Tipset::load_required(self, &TipsetKeys::from(self.roots()))
```

File: src/genesis/mod.rs [hunk 6]
```diff
@@ -119,7 +119,9 @@ where
         meta.sync()?;
     }
 
-    let ts = sm.chain_store().tipset_from_keys(&TipsetKeys::new(cids))?;
+    let ts = sm
+        .chain_store()
+        .tipset_from_keys(&TipsetKeys::new(cids.into()))?;
```
Comment (lemmih) [inline src/genesis/mod.rs:124]:
```suggestion
        .tipset_from_keys(&TipsetKeys::from(cids))?;
```

File: src/chain_sync/chain_muxer.rs (hunk 2 above)
Comment (lemmih) [reply src/chain_sync/chain_muxer.rs:378]: Oh, you already wrote it. Let's use it, then.

Comment (lemmih) [reply src/chain_sync/chain_muxer.rs:378]:
```suggestion
                let tipset_keys = TipsetKeys::from(request.heaviest_tip_set);
```

File: src/ipld/cid_vec.rs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        Self::default()
+    }
+
+    /// Creates a new `CidVec` from a single `Cid`, with the resulting variant of `CidVec` depending on the variant of the `Cid`.
+    pub fn new_from_cid(cid: Cid) -> Self {
+        match cid.try_into() {
+            Ok(CidVariant::V1DagCborBlake2b(bytes)) => Self::V1Cids(vec![bytes]),
+            _ => Self::AllCids(vec![cid]),
+        }
+    }
```
Comment (lemmih) [inline src/ipld/cid_vec.rs:88]: This can just be `From<Cid>`

File: src/ipld/cid_vec.rs [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    multihash::{self, Code::Blake2b256},
+    Cid,
+};
+use fvm_ipld_encoding::DAG_CBOR;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+// CidVec takes advantage of the fact that the V1 DAG-CBOR Blake2b-256 variant (which can be stored in 32 bytes vs 96 bytes for a `Cid` type) is +99.99% of all CIDs. CidVec defaults to the `Vec<[u8; BLAKE2B256_SIZE]>` type, only using the more expensive `Vec<Cid>` type when necessary.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum CidVec {
+    V1Cids(Vec<[u8; BLAKE2B256_SIZE]>),
```
Comment (lemmih) [inline src/ipld/cid_vec.rs:15]: They're not just V1. They're V1+DAG_CBOR+Blake2b.

File: src/libp2p/chain_exchange/provider.rs [hunk 9]
```diff
@@ -27,7 +27,7 @@ where
 
     loop {
         let mut tipset_bundle: TipsetBundle = TipsetBundle::default();
-        let tipset = match cs.tipset_from_keys(&TipsetKeys::new(curr_tipset_cids)) {
+        let tipset = match cs.tipset_from_keys(&TipsetKeys::new(curr_tipset_cids.into())) {
```
Comment (lemmih) [inline src/libp2p/chain_exchange/provider.rs:30]:
```suggestion
        let tipset = match cs.tipset_from_keys(&TipsetKeys::from(curr_tipset_cids)) {
```

File: src/message_pool/msgpool/test_provider.rs [hunk 10]
```diff
@@ -190,7 +190,7 @@ impl Provider for TestApi {
     fn load_tipset(&self, tsk: &TipsetKeys) -> Result<Arc<Tipset>, Error> {
         let inner = self.inner.lock();
         for ts in &inner.tipsets {
-            if tsk.cids == ts.cids() {
+            if tsk.cids == ts.cids().into() {
```
Comment (lemmih) [inline src/message_pool/msgpool/test_provider.rs:193]:
```suggestion
            if tsk == ts.key() {
```

File: src/rpc/sync_api.rs [hunk 11]
```diff
@@ -118,7 +118,7 @@ mod tests {
             let header = from_slice::<BlockHeader>(&bz).unwrap();
             let ts = Tipset::from(header);
             let db = cs_for_test.blockstore();
-            let tsk = ts.key().cids.clone();
+            let tsk = ts.key().cids.cids();
```
Comment (lemmih) [inline src/rpc/sync_api.rs:121]:
```suggestion
            let tsk = ts.key();
```

File: src/blocks/tipset.rs [hunk 12]
```diff
@@ -142,8 +143,8 @@ impl Tipset {
     pub fn load(store: impl Blockstore, tsk: &TipsetKeys) -> anyhow::Result<Option<Tipset>> {
         Ok(tsk
```
Comment (lemmih) [inline src/blocks/tipset.rs:144]: How about directly iterating over `tsk`?

**Review by lemmih — CHANGES_REQUESTED** [review]
Looks good. Just needs a bit of clean-up.

File: src/ipld/cid_vec.rs (hunk 8 above)
Comment (jdjaustin) [reply src/ipld/cid_vec.rs:15]: While `V1DAGCBORBlake2b256Cids` is more accurate, it seems a bit clunky. Do you suggest I go with that naming convention anyway, or do you have suggestions for a better name?

File: src/ipld/cid_vec.rs [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+use cid::{
+    multihash::{self, Code::Blake2b256},
+    Cid,
+};
+use fvm_ipld_encoding::DAG_CBOR;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+// CidVec takes advantage of the fact that the V1 DAG-CBOR Blake2b-256 variant (which can be stored in 32 bytes vs 96 bytes for a `Cid` type) is +99.99% of all CIDs. CidVec defaults to the `Vec<[u8; BLAKE2B256_SIZE]>` type, only using the more expensive `Vec<Cid>` type when necessary.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum CidVec {
```
Comment (aatifsyed) [inline src/ipld/cid_vec.rs:14]: If I've got a CidVec of length 1000, with just one rare CID, we're going to blow our optimisation.

- how common is this case?
- have we explored e.g
```rust
enum CidNiche {
    Small([u8; 32]),
    Heap(Box<Cid>)
}
```

We could even save a byte on a tagged pointer

File: src/ipld/cid_vec.rs [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+        }
+        cid_vec
+    }
+}
+
+impl From<&[Cid]> for CidVec {
+    fn from(vec: &[Cid]) -> Self {
+        vec.iter().cloned().collect()
+    }
+}
```
Comment (aatifsyed) [inline src/ipld/cid_vec.rs:68]: I don't think this should exist - users should control the clones, and there's a FromIterator implementation already

**Review by aatifsyed — CHANGES_REQUESTED** [review]

File: src/ipld/cid_vec.rs (hunk 13 above)
Comment (aatifsyed) [inline src/ipld/cid_vec.rs:14]: I also don't think this should be an `enum` - users can see an implementation detail

Comment (aatifsyed) [reply src/ipld/cid_vec.rs:14]: @lemmih I think I much prefer this:
```rust
use std::mem::size_of;

use cid::Cid;
use static_assertions::const_assert_eq;

#[repr(transparent)]
struct V1DagCborCid([u8; 32]);

const_assert_eq!(size_of::<Cid>(), 96);
const_assert_eq!(size_of::<V1DagCborCid>(), 32);

enum SmallCid {
    Stack(V1DagCborCid),
    Heap(Box<Cid>),
}

const_assert_eq!(size_of::<Box<Cid>>(), 8);
const_assert_eq!(size_of::<SmallCid>(), 40);
```

And we could use `Vec<SmallCid>` or even `SmallVec<SmallCid>` as required...

File: src/ipld/cid_vec.rs (hunk 7 above)
Comment (jdjaustin) [reply src/ipld/cid_vec.rs]: Replaced with `FrozenCids::from_iter`.

File: src/blocks/tipset.rs [hunk 15]
```diff
@@ -26,17 +27,17 @@ use super::{Block, BlockHeader, Error, Ticket};
...
+    pub fn new(cids: CidVec) -> Self {
         Self { cids }
     }
 
     /// Returns tipset header `cids`
-    pub fn cids(&self) -> &[Cid] {
-        &self.cids
+    pub fn cids(&self) -> Vec<Cid> {
+        self.cids.cids()
     }
```
[the remark this answers is not in the document]
Comment (jdjaustin) [reply src/blocks/tipset.rs]: Removed.

File: src/ipld/cid_vec.rs [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+                            .expect("failed to convert digest to CID"),
+                    )
+                })
+                .collect(),
+            Self::AllCids(cids) => cids.clone(),
+        }
+    }
+
+    /// Adds a CID to the `CidVec`, converting the `CidVec` to the `AllCids` variant if necessary.
+    pub fn push(&mut self, cid: Cid) {
```
[the remark this answers is not in the document]
Comment (jdjaustin) [reply src/ipld/cid_vec.rs]: No longer applicable per https://chainsafeio.slack.com/archives/C0144LHN732/p1691054770109249?thread_ts=1690999613.526739&cid=C0144LHN732.

File: src/ipld/cid_vec.rs (hunk 14 above)
Comment (jdjaustin) [reply src/ipld/cid_vec.rs]: Removed.

File: src/ipld/cid_vec.rs (hunk 13 above)
Comment (jdjaustin) [reply src/ipld/cid_vec.rs]: No longer applicable per https://chainsafeio.slack.com/archives/C0144LHN732/p1691054770109249?thread_ts=1690999613.526739&cid=C0144LHN732.

File: src/blocks/tipset.rs (hunk 12 above)
Comment (jdjaustin) [reply src/blocks/tipset.rs:145]: I believe the recent changes address this, but let me know if you were thinking of something else.

File: src/ipld/frozen_cids.rs [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+        for i in iter {
+            vec.push(i);
+        }
+        FrozenCids::from(vec)
+    }
+}
+
+// Converts `Vec<Cid>` to `FrozenCids(CidBox::V1DagCborBlake2bCids)` if possible; otherwise, converts to `FrozenCids(CidBox::AllCids)`.
+impl From<Vec<Cid>> for FrozenCids {
+    fn from(cids: Vec<Cid>) -> Self {
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:108]: This loses ordering, which I think is unexpected

**Review by lemmih — CHANGES_REQUESTED** [review]
The code is verbose, repeats itself, and has surprising inefficiencies. You either have to refactor it significantly or go the `Vec<SmallCid>` route.

File: src/ipld/frozen_cids.rs (hunk 17 above)
Comment (jdjaustin) [reply src/ipld/frozen_cids.rs:103]: Corrected.

File: src/ipld/cid_vec.rs (hunk 13 above)
Comment (jdjaustin) [reply src/ipld/cid_vec.rs]: Changed to `Vec<SmallCid>` based on requested changes below.

File: src/ipld/frozen_cids.rs [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+enum SmallCid {
+    Heap(Box<Cid>),
+    Inline([u8; 32]),
+}
+
+pub struct FrozenCidsIterator<'a> {
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:34]: I think this should wrap https://doc.rust-lang.org/stable/std/slice/struct.Iter.html, and inflate the inline CID on iteration using `From`

File: src/ipld/frozen_cids.rs [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+    }
+}
+
+impl FromIterator<Cid> for FrozenCids {
+    fn from_iter<T: IntoIterator<Item = Cid>>(iter: T) -> Self {
+        let mut vec = Vec::new();
+        for i in iter {
+            vec.push(i);
+        }
+        FrozenCids::from(vec)
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:98]:
```suggestion
        FrozenCids::from(iter.into_iter().collect::<Vec<_>>())
```

File: src/ipld/frozen_cids.rs [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+impl Default for FrozenCids {
+    fn default() -> Self {
+        FrozenCids(Box::new([]))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+enum SmallCid {
+    Heap(Box<Cid>),
+    Inline([u8; 32]),
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:31]:
```suggestion
    InlineDagCborV1([u8; 32]),
```

File: src/ipld/frozen_cids.rs [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+pub struct FrozenCids(Box<[SmallCid]>);
+
+impl Default for FrozenCids {
+    fn default() -> Self {
+        FrozenCids(Box::new([]))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+enum SmallCid {
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:29]:
```rust
impl From<Cid> for SmallCid { ... }
impl From<SmallCid> for Cid { ... }
impl Arbitrary for SmallCid { ... }
```

File: src/ipld/frozen_cids.rs [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+use fvm_ipld_encoding::DAG_CBOR;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+// Similar to the `CidHashMap` implementation, `FrozenCids` optimizes storage of
+// CIDs that would normally be stored as a vector of CIDs. The V1 DAG-CBOR
+// Blake2b-256 variant (which can be stored in 32 bytes vs 96 bytes for a `Cid`
+// type) is +99.99% of all CIDs, so very few CIDs need to be stored in the
+// `Heap(Box<Cid>)` variant of `SmallCid`. The Box type has been chosen to make
+// `FrozenCids` explicitly immutable due to the fact that the vectors are parsed
+// from CBOR and are immutable.
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:18]:
```suggestion
/// Similar to the `CidHashMap` implementation, `FrozenCids` optimizes storage of
/// CIDs that would normally be stored as a vector of CIDs. The V1 DAG-CBOR
/// Blake2b-256 variant (which can be stored in 32 bytes vs 96 bytes for a `Cid`
/// type) is +99.99% of all CIDs, so very few CIDs need to be stored in the
/// `Heap(Box<Cid>)` variant of `SmallCid`. 
/// 
/// We use `Box<[...]>` to save memory, avoiding vector overallocation.
```

File: src/ipld/frozen_cids.rs [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+        FrozenCids::from(vec)
+    }
+}
+
+impl From<Vec<Cid>> for FrozenCids {
+    fn from(cids: Vec<Cid>) -> Self {
+        let mut small_cids = Vec::with_capacity(cids.len());
+        for cid in cids {
+            match cid.try_into() {
+                Ok(CidVariant::V1DagCborBlake2b(bytes)) => small_cids.push(SmallCid::Inline(bytes)),
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:107]: `CidVariant` and `SmallCid` are very similar - can they be one type?

**Review by aatifsyed — CHANGES_REQUESTED** [review]

File: src/ipld/frozen_cids.rs (hunk 21 above)
Comment (jdjaustin) [reply src/ipld/frozen_cids.rs]: Changed `SmallCid` to `CidVariant` but implemented the above for `CidVariant`.

File: src/ipld/frozen_cids.rs (hunk 23 above)
Comment (jdjaustin) [reply src/ipld/frozen_cids.rs]: Changed `SmallCid` to `CidVariant` and modified `CidVariant` with generic variant.

File: src/ipld/frozen_cids.rs [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+
+impl Default for FrozenCids {
+    fn default() -> Self {
+        FrozenCids(Box::new([]))
+    }
+}
+
+pub struct IntoIter<'a> {
+    cids: std::slice::Iter<'a, CidVariant>,
+}
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:26]:
```suggestion
pub struct Iter<'a> {
    cids: std::slice::Iter<'a, CidVariant>,
}
```

File: src/ipld/frozen_cids.rs [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+
+impl<'de> Deserialize<'de> for FrozenCids {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(Self::from(<Vec<Cid>>::deserialize(deserializer)?))
+    }
+}
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:69]: I think you should implement `Serialize` and `Deserialize` for `CidVariant`, and then `#[derive(Serialize, Deserialize)]` for `FrozenCid`s

File: src/ipld/frozen_cids.rs [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+        cids
+    }
+}
+
+impl FrozenCids {
+    pub fn push(&self, cid: Cid) -> Self {
+        let mut cids = Vec::<Cid>::from(self);
+        cids.push(cid);
+        FrozenCids::from(cids)
+    }
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:118]: I don't think a `FrozenCids` struct should have a `push` method

File: src/ipld/frozen_cids.rs [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn contains(&self, cid: Cid) -> bool {
+        let cids = Vec::<Cid>::from(self);
+        cids.contains(&cid)
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:126]: Convert the `Cid` to a `CidVariant` and then call `contains`

File: src/ipld/frozen_cids.rs [hunk 28]
```diff
@@ -0,0 +1,176 @@
...
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use cid::multihash::{self, MultihashDigest};
+    use fvm_ipld_encoding::DAG_CBOR;
+    use quickcheck::Arbitrary;
+    use quickcheck_macros::quickcheck;
+
+    impl Arbitrary for FrozenCids {
```
Comment (aatifsyed) [inline src/ipld/frozen_cids.rs:138]: Can't this be derived?

File: src/utils/cid/mod.rs [hunk 29]
```diff
@@ -28,30 +28,47 @@ impl CidCborExt for Cid {}
...
+///
+/// Nearly all Filecoin CIDs are `V1`,`DagCbor` encoded, and hashed with `Blake2b256` (which has a hash
+/// length of 256 bits). Naively representing such a CID requires 96 bytes but `CidVariant` does it in
+/// only 40 bytes. If other types of CID become popular, they can be added to the `CidVariant`
+/// structure.
+///
+/// The `Generic` variant is used for CIDs that do not fit into the other variants.
+/// These variants are used for optimizing storage of CIDs in the `FrozenCids` structure.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub enum CidVariant {
```
Comment (aatifsyed) [inline src/utils/cid/mod.rs:43]:
```suggestion
pub enum SmallCid {
```

Comment (aatifsyed) [reply src/utils/cid/mod.rs:43]: should also `#[derive(Arbitrary)]`

File: src/utils/cid/mod.rs [hunk 30]
```diff
@@ -63,13 +80,27 @@ mod tests {
...
+            match bool::arbitrary(g) {
+                true => CidVariant::Generic(Box::new(Cid::arbitrary(g))),
+                false => CidVariant::from(Cid::new_v1(
+                    DAG_CBOR,
+                    multihash::Code::Blake2b256.digest(&u64::arbitrary(g).to_be_bytes()),
+                )),
+            }
+        }
+    }
+
```
Comment (aatifsyed) [inline src/utils/cid/mod.rs:103]: I don't see why we can't derive this

**Review by aatifsyed — APPROVED** [review]

File: src/ipld/frozen_cids.rs (hunk 24 above)
Comment (jdjaustin) [reply src/ipld/frozen_cids.rs]: Changed.

File: src/ipld/frozen_cids.rs (hunk 26 above)
Comment (jdjaustin) [reply src/ipld/frozen_cids.rs]: Good point, this was a leftover artifact from the previous iterations of this PR. Removed.

File: src/utils/cid/mod.rs (hunk 30 above)
Comment (jdjaustin) [reply src/utils/cid/mod.rs:103]: Two potential issues:
- what about the special logic to create a uniform distribution of the variants? (similar logic applies in the `FrozenCids` `Arbitrary` implementation)
- deriving `Arbitrary` for `CidVariant` raises an issue that ``the trait bound `[u8; 32]: Arbitrary` is not satisfied``

File: src/ipld/frozen_cids.rs (hunk 28 above)
Comment (jdjaustin) [reply src/ipld/frozen_cids.rs:132]: See https://github.com/ChainSafe/forest/pull/3275#discussion_r1288798090

File: src/utils/cid/mod.rs (hunk 29 above)
Comment (jdjaustin) [reply src/utils/cid/mod.rs:43]: I think `CidVariant` is more accurate--only the `V1DagCborBlake2b` variant is actually smaller, and renaming that variant to `SmallCid` would lose descriptiveness as well. For the `Arbitrary` discussion, see https://github.com/ChainSafe/forest/pull/3275#discussion_r1288798090.

File: src/ipld/frozen_cids.rs (hunk 25 above)
Comment (jdjaustin) [reply src/ipld/frozen_cids.rs]: Done.

File: src/utils/cid/mod.rs (hunk 30 above)
Comment (aatifsyed) [reply src/utils/cid/mod.rs:128]: `#[derive(Arbitrary)]` would result in a 50/50 distribution of the variants.

https://github.com/BurntSushi/quickcheck/pull/319, in the meantime you can use `#[arbitrary(gen(|g| std::array::from_fn(|_ix|Arbitrary::arbitrary(g))))]`

File: src/utils/cid/mod.rs (hunk 29 above)
Comment (aatifsyed) [reply src/utils/cid/mod.rs:44]: Not going to hold up the PR on this, but to explain my thinking.

`CidVariant` is a bit like `CidKind` - it sounds like a classification of a CID. Think of [`std::io::ErrorKind`](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html).
We don't want that - we're abstracting over different _storage methods_ of CIDs.

See other similar crates:
- https://crates.io/crates/smol_str
- https://docs.rs/smallvec/latest/smallvec/

Comment (aatifsyed) [reply src/utils/cid/mod.rs:44]: Also, I think this is the layout we want:
```rust
/// Save space on common CID types
pub struct SmallCid(SmallCidInner);

impl SmallCid {
    pub fn cid(&self) -> Cid { ... }
}

impl From<Cid> for SmallCid { ... }
impl From<SmallCid> for Cid { ... }

enum SmallCidInner {
    DagCborV1([u8; 32]),
    Heap(Box<Cid>)
}
```

Users shouldn't be able to see the enum variant

File: src/db/car/plain.rs [hunk 31]
```diff
@@ -152,7 +152,7 @@ impl<ReaderT: super::CarReader> PlainCar<ReaderT> {
     }
 
     pub fn heaviest_tipset(&self) -> anyhow::Result<Tipset> {
-        Tipset::load_required(self, &TipsetKeys::new(self.roots()))
+        Tipset::load_required(self, &TipsetKeys::new(self.roots().into()))
```
Comment (lemmih) [inline src/db/car/plain.rs:155]:
```suggestion
        Tipset::load_required(self, &TipsetKeys::from(self.roots()))
```

File: src/tool/subcommands/benchmark_cmd.rs [hunk 32]
```diff
@@ -183,7 +184,12 @@ async fn benchmark_exporting(
         compression_level,
         par_buffer(1024, blocks.map_err(anyhow::Error::from)),
     );
-    crate::db::car::forest::Encoder::write(&mut dest, ts.key().cids.clone(), frames).await?;
+    crate::db::car::forest::Encoder::write(
+        &mut dest,
+        Vec::<Cid>::from(ts.key().cids.clone()),
```
Comment (lemmih) [inline src/tool/subcommands/benchmark_cmd.rs:189]:
```suggestion
        Vec::<Cid>::from(&ts.key().cids),
```

File: src/utils/cid/mod.rs [hunk 33]
```diff
@@ -63,13 +105,27 @@ mod tests {
...
         Cid,
     };
     use fvm_ipld_encoding::DAG_CBOR;
+    use quickcheck::Arbitrary;
     use quickcheck_macros::quickcheck;
     use std::mem::size_of;
 
+    impl Arbitrary for CidVariant {
+        fn arbitrary(g: &mut quickcheck::Gen) -> Self {
+            // Quickcheck does not reliably generate the DAG_CBOR/Blake2b variant of V1 CIDs, but we can manually create them. This method will generate each variant of `CidVariant` with equal probability.
```
Comment (lemmih) [inline src/utils/cid/mod.rs:118]: This comment is not correct. It's only when generating a random `Cid` that it is unlikely to be a v1-dag_cbor-blake2b variant.

File: src/blocks/tipset.rs [hunk 34]
```diff
@@ -26,35 +27,36 @@ use super::{Block, BlockHeader, Error, Ticket};
...
-    pub fn cids(&self) -> &[Cid] {
-        &self.cids
-    }
-
     // Special encoding to match Lotus.
     pub fn cid(&self) -> anyhow::Result<Cid> {
         use fvm_ipld_encoding::RawBytes;
         let mut bytes = Vec::new();
-        for cid in self.cids() {
+        for cid in Vec::<Cid>::from(&self.cids) {
```
Comment (lemmih) [inline src/blocks/tipset.rs:42]:
```suggestion
        for cid in &self.cids {
```

File: src/chain/store/chain_store.rs [hunk 35]
```diff
@@ -247,9 +248,9 @@ where
         file.insert(*cid);
     }
 
-    pub fn unmark_block_as_validated(&self, cid: &Cid) {
+    pub fn unmark_block_as_validated(&self, cid: Cid) {
```
Comment (lemmih) [inline src/chain/store/chain_store.rs:251]: I'm not sure why this change is here.

File: src/chain_sync/tipset_syncer.rs [hunk 36]
```diff
@@ -882,7 +882,7 @@ async fn sync_headers_in_reverse<DB: Blockstore + Sync + Send + 'static, C: Cons
         }
         // Attempt to load the parent tipset from local store
         if let Ok(tipset) = chain_store.tipset_from_keys(oldest_parent.parents()) {
-            parent_blocks.extend_from_slice(tipset.cids());
+            parent_blocks.extend(&tipset.key().cids);
```
Comment (lemmih) [inline src/chain_sync/tipset_syncer.rs:885]:
```suggestion
            parent_blocks.extend(tipset.cids());
```

File: src/ipld/frozen_cids.rs [hunk 37]
```diff
@@ -0,0 +1,152 @@
...
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.cids.next() {
+            Some(cid) => match cid {
+                CidVariant::V1DagCborBlake2b(bytes) => {
+                    Some(Cid::from(CidVariant::V1DagCborBlake2b(*bytes)))
+                }
+                CidVariant::Generic(cid) => Some(*cid.clone()),
+            },
+            None => None,
+        }
```
Comment (lemmih) [inline src/ipld/frozen_cids.rs:49]:
```suggestion
        self.cids.next().map(Cid::from)
```

**Review by lemmih — CHANGES_REQUESTED** [review]
Getting closer.

File: src/chain/store/chain_store.rs (hunk 35 above)
Comment (jdjaustin) [reply src/chain/store/chain_store.rs:251]: Honestly I don't remember either, possibly a leftover artifact from one of the previous iterations of this PR.

Comment (jdjaustin) [reply src/chain/store/chain_store.rs]: Removed.

File: src/utils/cid/mod.rs (hunk 30 above)
Comment (jdjaustin) [reply src/utils/cid/mod.rs]: Thanks, was able to use this for now.

File: src/utils/cid/mod.rs (hunk 33 above)
Comment (jdjaustin) [reply src/utils/cid/mod.rs]: Changed to derive `Arbitrary`, so removed custom impl and associated comment.

**Review by lemmih — APPROVED** [review]
